src/bin/render-readmes.rs

```diff
@@ -0,0 +1,330 @@
+        }
+    }
+    if handle.response_code().unwrap() != 200 {
+        let response = String::from_utf8_lossy(&response);
+        println!(
+            "[{}-{}] Failed to get a 200 response: {}",
+            version.krate,
+            version.num,
+            response
+        );
+        return None;
+    }
+    let reader = Cursor::new(response);
+    let reader = GzDecoder::new(reader).expect(&format!(
+        "[{}-{}] Invalid gzip header",
+        version.krate,
+        version.num
+    ));
+    let mut archive = Archive::new(reader);
+    let mut entries = archive.entries().expect(&format!(
+        "[{}-{}] Invalid tar archive entries",
+        version.krate,
+        version.num
+    ));
+    let manifest: Manifest = {
+        let path = format!("{}-{}/Cargo.toml", version.krate, version.num);
+        let contents = find_file_by_path(&mut entries, Path::new(&path), &version).unwrap();
+        toml::from_str(&contents).expect(&format!(
+            "[{}-{}] Syntax error in manifest file",
+            version.krate,
+            version.num
+        ))
+    };
+    if manifest.package.readme.is_none() {
+        return None;
+    }
+    let rendered = {
+        let path = format!(
+            "{}-{}/{}",
+            version.krate,
+            version.num,
+            manifest.package.readme.unwrap()
+        );
+        let contents = find_file_by_path(&mut entries, Path::new(&path), &version).unwrap();
+        markdown_to_html(&contents).expect(&format!(
+            "[{}-{}] Couldn't render README",
+            version.krate,
+            version.num
+        ))
+    };
+    return Some(rendered);
+    #[derive(Deserialize)]
+    struct Package {
+        readme: Option<String>,
+    }
+    #[derive(Deserialize)]
+    struct Manifest {
+        package: Package,
+    }
+}
+
+/// Search an entry by its path in a Tar archive.
+fn find_file_by_path<R: Read>(
+    mut entries: &mut tar::Entries<R>,
+    path: &Path,
+    version: &EncodableVersion,
+) -> Option<String> {
+    let mut file = entries
+        .find(|entry| match *entry {
+            Err(_) => false,
+            Ok(ref file) => {
+                let filepath = match file.path() {
+                    Ok(p) => p,
+                    Err(_) => return false,
+                };
+                return filepath == path;
+            }
+        })
+        .expect(&format!(
+            "[{}-{}] file is not present: {}",
+            version.krate,
+            version.num,
+            path.display()
+        ));
+    match file {
+        Err(_) => None,
+        Ok(ref mut f) => {
+            let mut contents = String::new();
+            f.read_to_string(&mut contents).expect(&format!(
+                "[{}-{}] Couldn't read file contents",
+                version.krate,
+                version.num
+            ));
+            return Some(contents);
+        }
+    }
+}
+
+/// Creates and Arc over an App instance.
+fn make_app() -> Arc<App> {
```

Comment:
This duplication with `src/bin/server.rs` is a bit 😿😿😿

Do you think it would be possible to factor out the duplication?